### PR TITLE
Add benchmarking module for kNN evaluation of models

### LINKED ISSUE #77: Add callback for knn-validation
Add hooks for validation at the end of each epoch (only informative if there are labels). Let's make sure it can be switched off and that the user can pass a validation set of his choice.

### PR DESCRIPTION
**Pull request by IgorSusmelj**
### Changes

closes #77 

This PR adds a benchmarking module that can be used to quickly evaluate new SSL models on simple classification tasks using datasets such as CIFAR10, STL-10, ImageNet or others.

- Adds a `BenchmarkModule`. Inherit from it and pass it a PyTorch DataLoader to build up a feature bank to benchmark models.
- Add Docstrings and update documentation
- Update benchmarking code used for CIFAR10 benchmarks using the new module

I didn't add any unit tests since it's not a crucial part of the framework. However, I tested the implementation using our benchmark and did run it for a few epochs on a GPU as well as on CPU only. Everything seems to work fine.

### CODE REVIEW

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 1]
```diff
@@ -338,7 +243,7 @@ def configure_optimizers(self):
         for seed in range(n_runs):
             pl.seed_everything(seed)
             dataloader_train_ssl, dataloader_train_kNN, dataloader_test = get_data_loaders(batch_size)
-            model = Model(dataloader_train_kNN)
+            model = Model(dataloader_train_kNN, classes)
```
Comment (philippmwirth) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:246]: Didn't mention it the last time but you could name this `BenchmarkModel` to be a bit clearer.

File: lightly/utils/__init__.py [hunk 2]
```diff
@@ -13,3 +13,5 @@
 from lightly.utils.io import load_embeddings
 from lightly.utils.io import load_embeddings_as_dict
 from lightly.utils.embeddings_2d import fit_pca
+from lightly.utils.benchmarking import BenchmarkModule
+from lightly.utils.benchmarking import knn_predict
```
Comment (philippmwirth) [inline lightly/utils/__init__.py:17]: This is not necessary to do the benchmarks, right? It's already included in the `BenchmarkModule`. You just put it here if someone wants to use it on his own?

File: lightly/utils/benchmarking.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    def __init__(self,
+                 dataloader_kNN: DataLoader,
+                 num_classes: int,
+                 knn_k: int=200,
+                 knn_t: float=0.1):
+        super().__init__()
+        self.backbone = nn.Module()
+        self.max_accuracy = 0.0
+        self.dataloader_kNN = dataloader_kNN
+        #self.gpus = gpus
```
Comment (philippmwirth) [inline lightly/utils/benchmarking.py:159]: Why is this commented out?

**Review by philippmwirth — APPROVED** [review]
Very nice! Food for thought: What if our `SelfSupervisedEmbedding` or our `BaseEmbedding` class would inherit from the benchmarking module instead of the lightning module? We could still make the benchmarking voluntary. What do you think?

File: lightly/utils/__init__.py (hunk 2 above)
Comment (IgorSusmelj) [reply lightly/utils/__init__.py:17]: yes. In general, I don't like that we hide many useful methods within lightly. Having as much as possible private does not help anyone using lightly to build a custom model.

File: lightly/utils/benchmarking.py (hunk 3 above)
Comment (IgorSusmelj) [reply lightly/utils/benchmarking.py:159]: I removed it
